clang/include/clang/AST/Expr.h

```diff
@@ -4810,6 +4810,125 @@ class SourceLocExpr final : public Expr {
   friend class ASTStmtReader;
 };
 
+/// Represents a #embed "expression".
+class PPEmbedExpr final : public Expr {
+  SourceLocation BuiltinLoc, RParenLoc;
+  DeclContext *ParentContext;
+  StringLiteral *Filename = nullptr;
+  StringLiteral *BinaryData = nullptr;
+  IntegerLiteral *FakeChildNode = nullptr;
+  const ASTContext *Ctx = nullptr;
+
+public:
+  enum Action {
+    NotFound,
+    FoundOne,
+    Expanded,
+  };
+
+  PPEmbedExpr(const ASTContext &Ctx, StringLiteral *Filename,
+              StringLiteral *BinaryData, SourceLocation BLoc,
+              SourceLocation RParenLoc, DeclContext *Context);
+
+  /// Build an empty call expression.
+  explicit PPEmbedExpr(EmptyShell Empty) : Expr(SourceLocExprClass, Empty) {}
+
+  const DeclContext *getParentContext() const { return ParentContext; }
+  DeclContext *getParentContext() { return ParentContext; }
+
+  SourceLocation getLocation() const { return BuiltinLoc; }
+  SourceLocation getBeginLoc() const { return BuiltinLoc; }
+  SourceLocation getEndLoc() const { return RParenLoc; }
+
+  StringLiteral *getFilenameStringLiteral() const { return Filename; }
+  StringLiteral *getDataStringLiteral() const { return BinaryData; }
+
+  size_t getDataElementCount(ASTContext &Context) const;
+
+private:
+  template <bool Const>
+  class ChildElementIter
+      : public llvm::iterator_facade_base<
+            // FIXME: it seems reasonable to make this a random access iterator
+            // instead, but all current access patterns are a linear walk over
+            // the contents, so it's being left for follow-up work if needed.
+            ChildElementIter<Const>, std::input_iterator_tag,
+            std::conditional_t<Const, const IntegerLiteral *,
+                               IntegerLiteral *>> {
+    friend class PPEmbedExpr;
+
+    const ASTContext *Ctx = nullptr;
+    IntegerLiteral *FakeNode = nullptr;
+    StringRef DataRef;
+    LLVM_PREFERRED_TYPE(bool) unsigned long long IsSigned : 1;
```

Comment:
Is there a reason for not inserting a line break after `LLVM_PREFERRED_TYPE()`? That's how it usually being laid out in our code.

Reply:
Comment no longer applies because of a refactoring that removed the bit-field.